compiler/rustc_lint/src/builtin.rs

```diff
@@ -3098,3 +3094,54 @@ impl<'tcx> LateLintPass<'tcx> for DerefNullPtr {
         }
     }
 }
+
+declare_lint! {
+    /// The `const_transmute_ptr_to_number` lint detects transmutes from a
+    /// pointer to a number in a `const` context.
+    ///
+    ///
+    /// ### Example
+    ///
+    /// ```rust,no_run
+    /// # #![allow(dead_code)]
+    /// pub const fn x() -> usize {
+    ///     unsafe { std::mem::transmute(&0) }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Transmuting a pointer to a number may be [undefined behavior]. See
```

Comment:
```suggestion
    /// Transmuting a pointer to a number at compile-time is [undefined behavior]. See
```

Reply:
Is this concrete? I thought it was undecided.

Reply:
At compile-time I think it is pretty clear. The docs added in this PR explicitly says so
```
Transmuting pointers to integers in a `const` context is [undefined behavior][ub].
```

Reply:
Oh yeah 😆